### PR TITLE
Fix post-GC purging of streams view time histogram

### PR DESCRIPTION
**Pull request by emilk**
### What
* Closes https://github.com/rerun-io/rerun/issues/2517

The datastore is not the only place we store data. For each node in the entity tree we store a _time histogram_ of where we have data. This was never properly purged post GC - a very rough heuristic was instead used (throw away everything up to the oldest time in the store - which will be an even worse heursistic after https://github.com/rerun-io/rerun/pull/3357).

With this PR, the store gc will book-keep exactly what was deleted, and it will be properly purged from all secondary indices.

The resulting code is a bit complicated, because the store has no idea about the hierarchical nature of the entity paths, but we store the time histograms (and other book-keeping) for each node. That is, logging to `/foo/bar` we will note the data on `/`, `/foo` and `/foo/bar`, but the data will only be registered in the store for `/foo/bar`.

I also fixed a bunch of other smaller things that came up.

Best reviewed commit-by-commit.

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3364) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/3364)
- [Docs preview](https://rerun.io/preview/18cbc53fe96798cbe45ab84ddeb66aa183253e12/docs) 
- [Examples preview](https://rerun.io/preview/18cbc53fe96798cbe45ab84ddeb66aa183253e12/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: crates/re_data_store/src/entity_tree.rs [hunk 1]
```diff
@@ -301,45 +360,82 @@ impl EntityTree {
...
-            for columns in fields.values_mut() {
-                columns.purge(cutoff_times);
+            re_tracing::profile_scope!("ComponentStats");
+            for stats in components.values_mut() {
+                let ComponentStats {
+                    times,
+                    num_timeless_messages,
+                } = stats;
+
+                times.purge(path.hash(), deleted, deleted_by_us_and_children);
```
Comment (jleibs) [inline crates/re_data_store/src/entity_tree.rs:402]: As mentioned above, this purge should only happen if this component was actually cleared, which isn't guaranteed.

**Review by jleibs — APPROVED** [review]
~~Aside from the bug on the component-book-keeping~~ this looks good. ~~In reality that case is pretty rare, but still seems worth fixing for completeness.~~

File: crates/re_arrow_store/src/store_gc.rs [hunk 2]
```diff
@@ -58,6 +59,18 @@ impl std::fmt::Display for GarbageCollectionTarget {
     }
 }
 
+#[derive(Default)]
+pub struct Deleted {
+    /// What rows where deleted?
+    pub row_ids: HashSet<RowId>,
+
+    /// What time points where deleted for each entity?
+    pub timeful: IntMap<EntityPathHash, IntMap<Timeline, Vec<TimeInt>>>,
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_arrow_store/src/store_gc.rs:68]: I retract this retraction :facepalm:.

I was wrong about the storage structure, but think I'm still right about the underlying bug.  Multiple rows in the same table may contain different subsets of components for the same (timeline, timeint), so we should only be clearing the components that were contained within the row that is being cleared.

File: crates/re_arrow_store/src/store_gc.rs [hunk 3]
```diff
@@ -221,32 +231,39 @@ impl DataStore {
...
             // find all tables that could possibly contain this `RowId`
-            let temporal_tables = self.tables.iter_mut().filter_map(|((timeline, _), table)| {
-                timepoint.get(timeline).map(|time| (*time, table))
-            });
 
-            for (time, table) in temporal_tables {
-                num_bytes_to_drop -= table.try_drop_row(*row_id, time.as_i64()) as f64;
+            for ((timeline, ent_path_hash), table) in &mut self.tables {
+                if let Some(time) = timepoint.get(timeline) {
+                    num_bytes_to_drop -= table.try_drop_row(*row_id, time.as_i64()) as f64;
```
Comment (jleibs) [inline crates/re_arrow_store/src/store_gc.rs:240]: Fixing this is somewhat annoying.

The point where we currently have information about whether or not the component is cleared is down in:
https://github.com/rerun-io/rerun/blob/main/crates/re_arrow_store/src/store_gc.rs#L509-L511

Based on whether or not the row_index of that component is Some

**Comment by emilk** [comment]
Thanks for that @jleibs, it wasn't too bad keeping track of the individual components, and the code makes much more sense now imho!

**Pull request merged**